p2p/src/disconnection/p2p_disconnection_reducer.rs

```diff
@@ -22,27 +22,32 @@ impl P2pDisconnectedState {
 
         match action {
             P2pDisconnectionAction::Init { peer_id, reason } => {
-                let (dispatcher, state) = state_context.into_dispatcher_and_state();
-                let p2p_state: &P2pState = state.substate()?;
-
                 #[cfg(feature = "p2p-libp2p")]
                 if p2p_state.is_libp2p_peer(peer_id) {
-                    if let Some((addr, _)) = p2p_state
+                    if let Some((&addr, _)) = p2p_state
                         .network
                         .scheduler
                         .connections
                         .iter()
                         .find(|(_, conn_state)| conn_state.peer_id() == Some(peer_id))
                     {
+                        let Some(peer) = p2p_state.peers.get_mut(peer_id) else {
+                            bug_condition!("Invalid state for: `P2pDisconnectionAction::Finish`");
+                            return Ok(());
+                        };
+                        peer.status = P2pPeerStatus::Disconnecting { time: meta.time() };
+
+                        let dispatcher = state_context.into_dispatcher();
                         dispatcher.push(P2pNetworkSchedulerAction::Disconnect {
-                            addr: *addr,
+                            addr,
                             reason: reason.clone(),
                         });
                         dispatcher.push(P2pDisconnectionAction::Finish { peer_id: *peer_id });
```

Comment:
If it gets insta finished, without any event from mio, then disconnecting state seems useless.

Reply:
Let's keep it for further improvement. I will make `Finish` an effect of `Prune`, but not now.